File: sipivr.plugins/sipivr.speech/src/main/java/ru/sipivr/speech/core/DateSpeller.java
package ru.sipivr.speech.core;

import org.joda.time.LocalDate;
import ru.sipivr.speech.enums.Case;
import ru.sipivr.speech.enums.Gender;
import ru.sipivr.speech.enums.NumberType;

import java.util.HashMap;

/**
 * Created by Admin on 22.02.2016.
 */
public class DateSpeller {
    private NumberSpeller numberSpeller;

    private static final HashMap<Integer, String> MonthNames;

    static {
        MonthNames = new HashMap<>();

        MonthNames.put(1, "Jan");
        MonthNames.put(2, "Feb");
        MonthNames.put(3, "Mar");
        MonthNames.put(4, "Apr");
        MonthNames.put(5, "May");
        MonthNames.put(6, "Jun");
        MonthNames.put(7, "Jul");
        MonthNames.put(8, "Aug");
        MonthNames.put(9, "Sep");
        MonthNames.put(10, "Oct");
        MonthNames.put(11, "Nov");
        MonthNames.put(12, "Dec");
    }

    public DateSpeller(NumberSpeller numSpeller) {
        this.numberSpeller = numSpeller;
    }

    public CompositeLexicalUnit spell(LocalDate date) {
        return spell(date, Case.Nominative);
    }

    public CompositeLexicalUnit spell(LocalDate date, Case caseType) {
        CompositeLexicalUnit res = new CompositeLexicalUnit();

        NumberNoun dayOfMonth = new NumberNoun(date.getDayOfMonth());
        dayOfMonth.setNumberType(NumberType.Ordinal);
        dayOfMonth.setGender(Gender.Neuter);
        dayOfMonth.setCaseType(caseType);



      //  return Objects.equals(n.getValue(), "Year") ? String.format("Date/Year_%s", n.getCaseType().getValue()) : TranslateNoun(n);

        res.getChildren().add(numberSpeller.decomposeNumber(dayOfMonth));
        res.getChildren().add(new Noun("sipivr.speech.date", MonthNames.get(date.getMonthOfYear()), false));

        NumberNoun yearNumberNoun = new NumberNoun(date.getYear());
        yearNumberNoun.setCaseType(Case.Genitive);
        yearNumberNoun.setGender(Gender.Neuter);
        yearNumberNoun.setNumberType(NumberType.Ordinal);

        res.getChildren().add(numberSpeller.decomposeNumber(yearNumberNoun));

        res.getChildren().add(new Noun("sipivr.speech.date", "year_" + Case.Genitive.getValue(), false));

        return res;
    }
}

File: freeswitch/typescript/src/Utils/Speech/DateSpeller.ts
import { Noun } from "./Noun";
import { NumberNoun } from "./NumberNoun";
import { LexicalUnit } from "./LexicalUnit";
import { NumberSpeller } from "./NumberSpeller";
import { Case, NumberType, Gender } from "./Enums";

export class DateSpeller {
    private static MonthNames = [
        "Jan",
        "Feb",
        "Mar",
        "Apr",
        "May",
        "Jun",
        "Jul",
        "Aug",
        "Sep",
        "Oct",
        "Nov",
        "Dec",
    ]

    static spell(date: Date, caseType?: Case): LexicalUnit[] {
        if (caseType === undefined) {
            caseType = Case.Nominative;
        }

        let res: LexicalUnit[] = [];

        let dayOfMonth = new NumberNoun(date.getDate());
        dayOfMonth.NumberType = NumberType.Ordinal;
        dayOfMonth.Gender = Gender.Neuter;
        dayOfMonth.Case = caseType;
        //  return Objects.equals(n.getValue(), "Year") ? String.format("Date/Year_%s", n.getCaseType().getValue()) : TranslateNoun(n);

        NumberSpeller.decomposeNumber(dayOfMonth).forEach(f => res.push(f));
        res.push(new Noun(DateSpeller.MonthNames[date.getMonth()], "sipivr.date", false));

        let yearNumberNoun = new NumberNoun(date.getFullYear());
        yearNumberNoun.Case = Case.Genitive;
        yearNumberNoun.Gender = Gender.Neuter;
        yearNumberNoun.NumberType = NumberType.Ordinal;

        NumberSpeller.decomposeNumber(yearNumberNoun).forEach(f => res.push(f));

        res.push(new Noun("year_" + Case.Genitive, "sipivr.date", false));

        return res;
    }

    static spellWithoutYear(date: Date, caseType = Case.Nominative): LexicalUnit[] {
        let res: LexicalUnit[] = [];

        let dayOfMonth = new NumberNoun(date.getDate());
        dayOfMonth.NumberType = NumberType.Ordinal;
        dayOfMonth.Gender = Gender.Neuter;
        dayOfMonth.Case = caseType;
        //  return Objects.equals(n.getValue(), "Year") ? String.format("Date/Year_%s", n.getCaseType().getValue()) : TranslateNoun(n);

        NumberSpeller.decomposeNumber(dayOfMonth).forEach(f => res.push(f));
        res.push(new Noun(DateSpeller.MonthNames[date.getMonth()], "sipivr.date", false));

        return res;
    }

    static parse(text: string, format = "yyyy-MM-dd hh:mm:ss.fff") {
        var res = new Date(2008, 1, 1);
        var match = getMatch("yyyy");
        if (match !== null)
            res.setFullYear(match);
        else
            return null;
        
        match = getMatch("MM");
        if (match !== null)
            res.setMonth(match - 1);
        else
            return null;

        match = getMatch("dd");
        if (match !== null)
            res.setDate(match);
        else
            return null;

        match = getMatch("hh");
        if (match !== null)
            res.setHours(match);

        match = getMatch("mm");
        if (match !== null)
            res.setMinutes(match);

        match = getMatch("ss");
        if (match !== null)
            res.setSeconds(match);

        match = getMatch("fff");
        if (match !== null)
            res.setMilliseconds(match);

        return res;

        //----------------------------------------------------
        function getMatch(match: string): number {
            var pos = format.search(match);
            return (pos != -1) ? parseInt(text.substr(pos, match.length), 10) : null;
        }
    }
}